AI #1838: Add TagDetails column to CostAndUsage (#1904)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/tagdetails.md

# Tag Details

The Tag Details column is a superset of [Tags](#tags) which includes additional properties describing tag eligibility and tag provenance from all [*tag sources*](#glossary:tag-source). Tag Details can be used to determine whether a [*charge*](#glossary:charge) was eligible to be tagged, improving the accuracy of tag coverage calculations. Tag Details can also be used to determine the origin of the [*finalized tag*](#glossary:finalized-tag) as well as find tag values present on ancestor sources.

## Requirements

### Column Requirements

The TagDetails column adheres to the following requirements:

> * TagDetails SHOULD be present in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset) when the data generator supports setting user or provider-defined tags.

Reviewer: I agree with SHOULD here, we've heard lots of feedback from customers that they'd like to be in control of when columns are populated that would drastically increase the size of the dataset (see AWS Split Cost Allocation in Legacy CUR as an example). If this was a MUST, then practitioners would be getting 3-7x larger reports by default, which seems unnecessary to force.

Reviewer 2: `TagDetails SHOULD be present in a Cost and Usage`

should be at the dataset Level

Author: Yeah, I would not have supported this column if #1091 wasn't included in this release. I personally view this column as very useful in a limited context, but certainly not necessary as part your daily ingestion jobs.

Author: Moved requirement to dataset level

Author: Justin, I added to that the inclusion requirements that this column SHOULD NOT be included if dataset configuration (column selection in export) is not supported. I believe this covers the concerns you and I had around this.

* TagDetails MUST be of type String.
* TagDetails MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* TagDetails MUST conform to [JsonObjectFormat](#attributes.jsonobjectformat) requirements.
* TagDetails nullability is defined as follows:

>   * TagDetails MUST be null when Tags is null.

Reviewer 3: Is this correct? If Tags is null but the resource was eligible for tagging, then we would want to know how with `UntaggedSources` in `TagDetails`, no?

Author: I agree. Irena and I spoke about this during last week's meeting and I convinced her of the need to change some prior guidance/decision. Will be correcting today.

Reviewer 4: This is still open and out there, so I'm going to add my two cents: 

I've specifically encountered scenarios with containers and VMs where tags weren't available, but metadata associated with untagged sources was still critical for cost identification. In those cases, 
having access to that metadata — even in the absence of finalized tags — was the only way to attribute costs meaningfully. Requiring TagDetails to be null whenever Tags is null would silently discard exactly that kind of useful context. 

I'd suggest changing to: 
`* TagDetails MAY be null when Tags is null.`

Author: Yeah, the comment was still open but it was hidden because it was outdated. Here's what's currently reflected for nullability. 
  * TagDetails MUST NOT be null unless all of the following are true:
    * Tags column is null.
    * Tags are not present in any other *tag sources*.
    * No *tag sources* are supported for any user-defined tag scheme.

I even included an example that shows the Tags column is empty, but the TagDetails has ancestor tagged values present.

### Object Schema Requirements

Tag Details consists of a valid JSON object which contains objects for one or more Tags schemes which each contain an object describing the tags as well as an array of eligible *tag sources* which do not contain tags.

When TagDetails is not null, the JsonObjectFormat for TagDetails adheres to the following requirements:

* TagDetails root object MUST contain an object for each tag scheme present in Tags.
  * The key for each tag scheme object MUST align to the tag prefix (up to but not including the forward slash) used to define the tag scheme in Tags.
  * The key for the unprefixed user-defined tag scheme in Tags MUST be "Default".
* Each tag scheme object MUST contain an object for each tag key used in that tag scheme.
  * Each tag key object MUST contain FOCUS-defined tag properties.
  * FOCUS-defined tag properties are subject to the additional requirements:
    * Tag property key MUST match the spelling and casing specified for the FOCUS-defined property.
    * Tag property value MUST be of the type specified for that property.
    * Tag properties MUST adhere to additional normative requirements specific to that property.

>   * Each tag key object MUST contain an object with they key "AncestorTaggedSources".

Reviewer: ```suggestion
  * Each tag key object MUST contain an object with the key "AncestorTaggedSources".
```

    * The value of AncestorTaggedSources MUST be null when the tag key from the corresponding tag key object within the tag scheme is not present in any *tag source* other than the *tag source* which results in the *finalized tag*.

>     * When the value of AncestorTaggedSources is not null, each object in AncestorTaggedSources MUST have a key denoting the *tag source*.

Reviewer 5: This requirement implies an Array structure instead of the Map structure shown in your example.

Your current example farther down follows the Map pattern:

```json
"AncestorTaggedSources": {
  "Subscription": {              // The label IS the Source Type
    "TagValue": "foo"
  },
  "Resource Group": {
    "TagValue": "bar"
  }
}
```

Whereas this normative requirement implies an Array structure like:

```json
"AncestorTaggedSources": [
  {
    "SourceType": "Subscription",   // You have to open the object to find out what it is
    "TagValue": "foo"
  },
  {
    "SourceType": "Resource Group",
    "TagValue": "bar"
  }
]
```

To reconcile this, I think we change the normative requirement to match your example.  Something like:

```suggestion
    * AncestorTaggedSources MUST be a JSON object where each key represents the tag source type (e.g., 'Subscription') and the value is an object containing the tag properties.
```

Author: I see the confusion. You are correct that these should be key-value pairs with the key being the ancestor tag source and the value being the object. The object will contain at least the property (key-value pair) TagValue.

I think this will be much easier to write clearly and concisely after Mike's revision to how we're structuring these. Like I said in one of the meetings, I was having difficulties finding the right words to clearly describe (and differentiate) the situations, including {tags (the concept) and "Tags": the object} and {key of the tag (concept) and key inside of the json object}.

If you look at how I've written the rest of this, the way I've been referring to these is 
"ObjectKey":{
  "PropertyKey":"PropertyValue"
  } 
with the entire thing referred to as the _ObjectKey_ Object

So I agree that a revision is necessary, but right now the way this is written appears to be consistent with the way the rest of the PR is written.

Author: I believe this is more clear in #2078

* TagValue MUST have the value of true (boolean) when the TagScheme does not support values.
* Data generator MUST NOT alter tag values unless applying true (boolean) to valueless tags.
* TagValue MAY be null when the the data generator supports setting a null value for a key-value pair type tag.

<b>TagSource</b>

"TagSource" denotes the type of *tag source* where the tag key is present.

The "TagSource" property adheres to the following requirements:

* TagSource MUST be present in each tag key object in the Tags object.
* TagSource MUST be present in each tag source object in the AncestorTaggedSources object.
* TagSourceId MUST contain the type of *tag source* where the tag key is present.

<b>TagSourceId</b>

"TagSourceId" denotes the identifier of the specific *tag source* where the tag key is present.

The "TagSourceId" property adheres to the following requirements:

* TagSourceId MUST be present in each tag key object in the Tags object.
* TagSourceId MUST be present in each tag source object in the AncestorTaggedSources object.
* TagSourceId MUST contain the identifier of the TagSource.

## Overview

> ### Object Scheme

Reviewer 3: ```suggestion
### Object Schema
```
Checking the use of schema vs scheme. Ditto for line 84.

Author: Thanks, I called this out as one of the things to watch for as I was writing (and re-re-writing) it many times.

This is schema (the shape of the data). 
Line 84 should be tag scheme (the provider defined method of tagging)

| _TagScheme_ | _root object_ | Object | TRUE | One or more objects containing information about a prefixed tag scheme in the Tags column. |
| Tags | _tag scheme object_ | Object | TRUE | An object containing all tag keys present in the Tags column. |
| UntaggedSources | _tag scheme object_ | Array | TRUE | A list of sources which support this tag scheme (i.e. are taggable) that had no tags applied. |
| _TagKey_ | Tags | Object | TRUE | An object containing the properties of the *finalized tag* as well as other *tag sources* where this key was present. |
| AncestorTaggedSources | _TagKey_ | Object | TRUE | An object containing all *tag sources* where the corresponding tag key was present which did not result in the *finalized tag*. |
| _AncestorTaggedSource_ | AncestorTaggedSources | Object | Conditional | An object containing the the properties of the tag present in the *tag source*. |

### Object Entries

The tag key object and tag source objects contain the following properties:

| Key | ValueType | Required | Description |
| ----- | ---- | ---------- | ----------- |

> | TagValue | String | TRUE |  The value of the tag key. |

Reviewer 5: You've got this listed here as a String -- but line 56 is currently:

> * TagValue MUST have the value of true (boolean) when the TagScheme does not support values.

Because Tags also allows String and Boolean, you likely need to allow similar here.

Author: Resolved in the current version 
| TagValue | String, Boolean, or Number | TRUE |  The value associated with the tag key. |

| TagSource | String | TRUE | The type of *tag source* where the tag key is present. |
| TagSourceId | String | TRUE | The identifier of the *tag source* where the tag key is present. |

### Example

```json
{
  "Default": {
    "Tags": {
      "foo": {

>         "FocusColumn": "ResourceId",

Reviewer 5: What is `FocusColumn`?  It's here several times in the example, but it's not anywhere in the requirements.

Author: Yeah, this is an open item for discussion this week. It was included in the original proposal but as I was writing all of this out, I was conflicted.

Author: This has been removed

# Tag Details

The Tag Details column is a superset of [Tags](#tags) which includes additional properties describing tag eligibility and tag provenance from all [*tag sources*](#glossary:tag-source). Tag Details can be used to determine whether a [*charge*](#glossary:charge) was eligible to be tagged, improving the accuracy of tag coverage calculations. Tag Details can also be used to determine the origin of the [*finalized tag*](#glossary:finalized-tag) as well as find tag values present on ancestor sources which do not appear in the Tags column.

## Requirements

### Column Requirements

The TagDetails column adheres to the following requirements:

* TagDetails MUST be of type String.
* TagDetails MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* TagDetails MUST conform to [JsonObjectFormat](#attributes.jsonobjectformat) requirements.
* TagDetails nullability is defined as follows:
  * TagDetails MUST NOT be null unless all of the following are true:
    * Tags column is null.
    * Tags are not present in any other *tag sources*.

>     * No *tag sources* are supported for any user-defined or provider-defined tag scheme.

Author: ```suggestion
    * No *tag sources* are supported for any user-defined tag scheme.
```

# Tag Details

The Tag Details column is a superset of [Tags](#tags) which includes additional properties describing tag eligibility and tag provenance from all [*tag sources*](#glossary:tag-source). Tag Details can be used to determine whether a [*charge*](#glossary:charge) was eligible to be tagged, improving the accuracy of tag coverage calculations. Tag Details can also be used to determine the origin of the [*finalized tag*](#glossary:finalized-tag) as well as find tag values present on ancestor sources which do not appear in the Tags column.

## Requirements

### Column Requirements

The TagDetails column adheres to the following requirements:

* TagDetails MUST be of type String.
* TagDetails MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* TagDetails MUST conform to [JsonObjectFormat](#attributes.jsonobjectformat) requirements.
* TagDetails nullability is defined as follows:
  * TagDetails MUST NOT be null unless all of the following are true:
    * Tags column is null.
    * Tags are not present in any other *tag sources*.

>     * No *tag sources* are supported for any user-defined tag scheme.

Author: ```suggestion
    * No *tag sources* are supported for any user-defined [*tag scheme*](glossary:tag-scheme).
```

### Object Schema Requirements

> Tag Details consists of a valid JSON object which contains objects for one or more Tags schemes which each contain an object describing the tags as well as an array of eligible *tag sources* which do not contain tags.

Reviewer 6: Non-blocking: "one or more Tags schemes" could be read as referencing the Tags column name. Consider "one or more tag schemes" (lowercase) or linking the new glossary term: "[*tag schemes*](#glossary:tag-scheme)."

Author: I added a definition for tag schemes to the glossary whereas previously, it had only really been used in the Tags column definition.

Author: ```suggestion
Tag Details consists of a valid JSON object which contains objects for one or more *tag schemes* which each contain an object describing the tags as well as an array of eligible *tag sources* which do not contain tags.
```

When TagDetails is not null, the JsonObjectFormat for TagDetails adheres to the following requirements:

* TagDetails root object MUST contain an object for each tag scheme present in Tags.
  * The key for each tag scheme object MUST align to the tag prefix (up to but not including the forward slash) used to define the tag scheme in Tags.
  * The key for the unprefixed user-defined tag scheme in Tags MUST be "Default".

> * Each tag scheme object MUST contain an object for each tag key used in that tag scheme.

Reviewer 6: Non-blocking: "each tag key used in that tag scheme" is ambiguous. In the ancestor-only example (lines 330-348), tag key "foo" appears in the TagDetails Tags object despite not being in the Tags column. "Used" could mean "present in any tag source for this charge" or "present in the Tags column." The ancestor-only pattern implies the former, but implementors could reasonably read it either way.

Classified as non-blocking because the fix requires a design decision on what "used" means, not a text replacement.

Author: Updated to "each tag key applied to any tag source in that tag scheme"

  * Each tag key object MUST contain FOCUS-defined tag properties.
  * FOCUS-defined tag properties are subject to the additional requirements:
    * Tag property key MUST match the spelling and casing specified for the FOCUS-defined property.
    * Tag property value MUST be of the type specified for that property.
    * Tag properties MUST adhere to additional normative requirements specific to that property.
  * Each tag key object MUST contain an object with the key "AncestorTaggedSources".
    * The value of AncestorTaggedSources MUST be null when the tag key from the corresponding tag key object within the tag scheme is not present in any *tag source* other than the *tag source* which results in the *finalized tag*.
    * When the value of AncestorTaggedSources is not null, each object in AncestorTaggedSources MUST have a key denoting the *tag source*.
      * Each tag source object MUST contain FOCUS-defined tag properties.
    * AncestorTaggedSources SHOULD contain objects for *tag sources* which did not result in the *finalized tag*.
* Each tag scheme object MUST contain an array with the key "UntaggedSources".

>   * Untagged Sources array MUST contain all *tag sources* for the corresponding tag scheme which are eligible to be tagged for the *charge*.

Reviewer 6: ```suggestion
  * UntaggedSources array MUST contain all *tag sources* for the corresponding tag scheme which are eligible to be tagged for the *charge* but have no tags applied.
```

Current text reads as "all sources eligible to be tagged" which would include tagged sources. Adding "but have no tags applied" matches the description in the Object Schema table (line 92: "sources which support this tag scheme...that had no tags applied"). Also normalizes the key name from "Untagged Sources" (with space) to the exact key "UntaggedSources."

>   * The value of Untagged Sources MUST be null when there are no eligible *tag sources* which contain no tags.

Reviewer 6: ```suggestion
  * The value of UntaggedSources MUST be null when there are no eligible *tag sources* which contain no tags.
```

Same key-name normalization: "Untagged Sources" to "UntaggedSources."

### Content Requirements

The following keys are used for tag properties to facilitate standardized extraction of data across providers. FOCUS-defined keys will appear in the list below and data generator-defined keys will be prefixed with "x_" to make them easy to identify as well as prevent collisions.

<b>TagValue</b>

"TagValue" represents the tag value associated with the tag key for the corresponding tag scheme and *tag source*.

The "TagValue" property adheres to the following requirements:

* TagValue MUST be present in each tag key object in the Tags object.
* When TagValue is directly contained in the tag key object, TagValue MUST represent the *finalized tag*.
* TagValue MUST be present in each tag source object in the AncestorTaggedSources object.
* TagValue MUST have the value of true (boolean) when the TagScheme does not support values.
* Data generator MUST NOT alter tag values unless applying true (boolean) to valueless tags.

> * TagValue MAY be null when the the data generator supports setting a null value for a key-value pair type tag.

Reviewer 6: ```suggestion
* TagValue MAY be null when the data generator supports setting a null value for a key-value pair type tag.
```

Typo: "the the data generator."

<b>TagSource</b>

"TagSource" denotes the type of *tag source* where the tag key is present.

The "TagSource" property adheres to the following requirements:

* TagSource MUST be present in each tag key object in the Tags object.
* TagSource MUST contain the type of *tag source* where the tag key is present.

> * TagSource SHOULD NOT be present in each tag source object in the AncestorTaggedSources object.

Reviewer 6: Non-blocking: TagSource is redundant in AncestorTaggedSources because the key itself denotes the source type. If there's no valid reason a generator would include it, MUST NOT might be cleaner than SHOULD NOT.

Author: Agreed. Cleaned up and moved.

* TagSourceId MUST contain the identifier of the TagSource.

## Overview

### Object Schema

The object contains an object for each tag scheme present in the Tags column.

| Key | Parent | ValueType | Required | Description |
| ----- | ---- | ---- | ---------- | ----------- |
| Default | _root object_ | Object | TRUE | The object containing information about the unprefixed tag scheme in the Tags column. |
| _TagScheme_ | _root object_ | Object | TRUE | One or more objects containing information about a prefixed tag scheme in the Tags column. |
| Tags | _tag scheme object_ | Object | TRUE | An object containing all tag keys present in the Tags column. |
| UntaggedSources | _tag scheme object_ | Array | TRUE | A list of sources which support this tag scheme (i.e. are taggable) that had no tags applied. |
| _TagKey_ | Tags | Object | TRUE | An object containing the properties of the *finalized tag* as well as other *tag sources* where this key was present. |
| AncestorTaggedSources | _TagKey_ | Object | TRUE | An object containing all *tag sources* where the corresponding tag key was present which did not result in the *finalized tag*. |

> | _AncestorTaggedSource_ | AncestorTaggedSources | Object | Conditional | An object containing the the properties of the tag present in the *tag source*. |

Reviewer 6: ```suggestion
| _AncestorTaggedSource_ | AncestorTaggedSources | Object | Conditional | An object containing the properties of the tag present in the *tag source*. |
```

Typo: "the the properties."

### Object Entries

The tag key object and tag source objects contain the following properties:

| Key | ValueType | Required | Description |
| ----- | ---- | ---------- | ----------- |

> | TagSource | String | Conditional | The type of *tag source* where the tag key is present. |

Author: This should not be conditional. It MUST be present in tag key object. 
However, it acts as the ObjectKey for the AncestorTaggedSource object so it does not need to be included in the properties of that object.

This will be easier to rewrite once we change the structure

Author: This is resolved in #2078

| TagSourceId | String | TRUE | The identifier of the *tag source* where the tag key is present. |

> | TagValue | String or Boolean | TRUE |  The value associated with the tag key. |

Reviewer 6: ```suggestion
| TagValue | String, Boolean, or Number | TRUE |  The value associated with the tag key. |
```

The JSON schema (lines 252-265) allows `number` type and the second example scenario uses `12.2` (line 431). The table should match.

Worth noting: Number-typed tag values have schema and example support but no normative requirement text in the Content Requirements section (lines 54-59). The TF may want explicit guidance on when Number values appear.

Author: Yep, I had removed the option for Number because I could not think of a case in which number could be used. If I set a tag as product = 8675309, it would still show in tags as {"product":"8675309"} because I've put a number into a string field since the tag scheme supports strings. The definition of Tags states that the provider can't change the value (except TRUE for key-only tags), so it could never be a number. However, it later occurred to me that ProviderDefined tags could be numeric because they can define them on a per-key basis however they want.

          "anyOf": [
            {
              "type": "string"
            },
            {
              "type": "boolean"
            },
            {
              "type": "number"
            },
            {
              "type": "null"
            }
          ]
        }
      },
      "required": [
        "TagSourceId",
        "TagValue"
      ]
    }
  }
}
```

## Example Scenarios

### Ancestor tags only

Details:
* One Tag Scheme supported (default).
* Charge is for API usage which has no resource to tag, but this charge does support job tagging in the API request payload.
  * No tags were applied in the API request payload.
* One tag was applied at the ancestor level.

```json
{
  "Default": {
    "Tags": {
      "foo": {
        "TagSource": null,
        "TagSourceId": null,
        "TagValue": null,
        "AncestorTaggedSources": {
          "Project": {
            "TagSourceId": "gcp-project-8675309",
            "TagValue": "bar"
          }
        }
      }
    },
    "UntaggedSources": ["api-job-label"]
  }
}
```

The corresponding Tags column would either be null or contain an empty object:

```json
{}
```

> ### Tags from multiple sources and schemas with different value types

Reviewer 6: ```suggestion
### Tags from multiple sources and schemes with different value types
```

Per the new glossary term [*tag scheme*](#glossary:tag-scheme). "Schemas" refers to data shape; "schemes" refers to tagging concepts.

Author: Yes. This was on my BOLO list
